src/importation/index.spec.js

```diff
@@ -0,0 +1,184 @@
+      const finishedImportation = await finishedImportationPromise;
+      expect(finishedImportation).toEqual({
+        file,
+        report: {
+          imported: 1,
+          skipped: [],
+          total: 1,
+          unsaved: 0
+        },
+        status: Status.COMPLETE_SUCCESS
+      });
+    });
+
+    it("notifies of progress & saves contacts", async () => {
+      const file = fixture.file("two-contacts.vcf");
+      const importation = { file, status: Status.READY };
+      const save = jest.fn().mockResolvedValue();
+      const onProgress = jest.fn();
+      await Importation.run(importation, { save, onProgress })
+        .finishedImportationPromise;
+      expect(save).toBeCalledWith(
+        expect.objectContaining({ fullname: "Contact 1" })
+      );
+      expect(save).toBeCalledWith(
+        expect.objectContaining({ fullname: "Contact 2" })
+      );
+      expect(onProgress.mock.calls).toEqual([
+        [{ current: 0, total: 2 }],
+        [{ current: 1, total: 2 }],
+        [{ current: 2, total: 2 }]
+      ]);
+    });
+
+    it("only notifies of progress when no contact", async () => {
+      const file = fixture.file("broken/end-vcard-missing.vcf");
+      const importation = { file, status: Status.READY };
+      const save = jest.fn().mockResolvedValue();
+      const onProgress = jest.fn();
+      await Importation.run(importation, { save, onProgress })
+        .finishedImportationPromise;
+      expect(save).not.toBeCalled();
+      expect(onProgress.mock.calls).toEqual([[{ current: 0, total: 0 }]]);
+    });
+  });
+
+  describe("canRun", () => {
+    const { canRun, init, run, selectFile } = Importation;
+    const { validFile, invalidFile } = fixture;
+
+    it("is false initially", () => {
+      expect(canRun(init())).toBe(false);
+    });
+
+    it("is false when invalid file selected", () => {
+      expect(canRun(selectFile(invalidFile(), init()))).toBe(false);
+    });
+
+    it("is true when valid file selected", () => {
+      expect(canRun(selectFile(validFile(), init()))).toBe(true);
+    });
+
+    it("is false when running", async () => {
+      const { runningImportation, finishedImportationPromise } = run(
+        selectFile(validFile(), init())
+      );
+      expect(canRun(runningImportation)).toBe(false);
+      await finishedImportationPromise;
+    });
+
+    describe("when already run", () => {
+      it("is true when importation can be retried – see canRetry() below", async () => {
+        const importation = await run(selectFile(validFile(), init()))
+          .finishedImportationPromise;
+        const expectCanRun = (bool, reportFields) => {
+          const report = {
+            ...importation.report,
+            ...reportFields
+          };
+          const status = Status.fromReport(report);
+          expect(canRun({ ...importation, report, status })).toBe(bool);
+        };
+        expectCanRun(false, { total: 0, imported: 0, unsaved: 0 });
```

Comment:
```js
const expectCanRun = reportFileds => bool => {
  const report = {
    ...importation.report,
    ...reportFields
  }
  const status = Status.fromReport(report)
  return expect(canRun({ ...importation, report, status })).toBe(bool)
}
```

becomes

```js
const expectCanRun = reportFileds => {
  return bool => {
    const report = {
      ...importation.report,
      ...reportFields
    }
    const status = Status.fromReport(report)
    return expect(canRun({ ...importation, report, status })).toBe(bool)
  }
}
```

becomes

```js
const expectCanRun = reportFileds => {
  const report = {
    ...importation.report
    ...reportFields
  }
  const status = Status.fromReport(report)
  return bool => {
    return expect(canRun({ ...importation, report, status })).toBe(bool)
  }
}
```

becomes


```js
const expectCanRun = reportFileds => {
  const report = {
    ...importation.report
    ...reportFields
  }
  const status = Status.fromReport(report)
  return expect(canRun({ ...importation, report, status }))
}
```

Then you can call the function like this

```js
expectCanRun({ total: 1, imported: 0, unsaved: 0 }).toBe(false)
```
that seems to me more readable.

Reply:
👌